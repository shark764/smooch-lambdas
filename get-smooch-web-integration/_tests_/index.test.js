const { validateTenantPermissions } = require('serenova-js-utils/lambda/api');

jest.mock('smooch-core/lib/api/integrations');
jest.mock('serenova-js-utils/lambda/api');
jest.mock('aws-sdk');
jest.mock('smooch-core/lib/smooch');

const mockGetSecretValue = jest.fn(() => {})
  .mockImplementation(() => ({
    promise: () => ({
      SecretString: JSON.stringify({
        SecretId: 'us-east-1-dev-smooch',
      }),
    }),
  }));

const mockGet = jest.fn()
  .mockImplementation(() => ({
    promise: () => ({
      Item: [{ 'app-id': '5e31c81640a22c000f5d7f28', type: '' }],
    }),
  }));

validateTenantPermissions.mockReturnValue(true);

const event = {
  params: {
    'tenant-id': '66d83870-30df-4a3b-8801-59edff162034',
    id: '667802d8-2260-436c-958a-2ee0f71f73f0',
  },
  identity: {
    'user-id': '667802d8-2260-436c-958a-2ee0f71f73f0',
  },
};

jest.mock('aws-sdk', () => ({
  config: {
    update: jest.fn().mockImplementation(),
  },
  SecretsManager: jest.fn().mockImplementation(() => ({ getSecretValue: mockGetSecretValue })),
  DynamoDB: {
    DocumentClient: jest.fn().mockImplementation(() => ({ get: mockGet })),
  },
}));

const mockGetIntegrations = jest.fn(() => ({
  integration: {
    brandColor: '#1b3de5',
    conversationColor: '#e53f1b',
    actionColor: '#1be54c',
    prechatCapture: {
      fields: [{ name: 'smooch' }],
    },
    originWhitelist: ['url1', 'url2'],
    whitelistedUrls: [],
    integrationOrder: '',
    _id: '667802d8-2260-436c-958a-2ee0f71f73f0',
    displayName: 'smooch',
    status: 'done',
    type: 'web',
  },
}));

jest.mock('smooch-core', () => jest.fn(() => ({
  integrations: {
    get: mockGetIntegrations,
  },
})));

beforeAll(() => {
  global.process.env = {
    AWS_REGION: 'us-east-1',
    ENVIRONMENT: 'dev',
    smooch_api_url: 'mock-smooch-api-url',
  };
});

const { handler } = require('../index');

describe('get-smooch-web-integration', () => {
  describe('Everything is successful', () => {
    let result;
    beforeAll(async () => {
      result = await handler(event);
    });
    it('sends back status 200 if the code runs without error', async () => {
      expect(result).toMatchSnapshot();
    });

    it('sends back arguments with which validateTenantPermissions was called', async () => {
      expect(validateTenantPermissions.mock.calls).toMatchSnapshot();
    });

    it('sends back arguments with which mockGetSecretValue was called', async () => {
      expect(mockGetSecretValue.mock.calls).toMatchSnapshot();
    });

    it('sends back arguments with which mockGetIntegrations was called', async () => {
      expect(mockGetIntegrations.mock.calls).toMatchSnapshot();
    });
  });

  it('sends back a 400 error when there are invalid parameters', async () => {
    const mockevent = {
      params: {
        'tenant-id': '66d83870-30df-4a3b-8801-59edff162034',
      },
      identity: {
        'user-id': '667802d8-2260-436c-958a-2ee0f71f73f0',
      },
    };
    const result = await handler(mockevent);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 400 error when there are not enough permissions', async () => {
    validateTenantPermissions.mockReturnValueOnce(false);
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 500 when there is a error retrieving digital channels credentials ', async () => {
    mockGetSecretValue.mockRejectedValueOnce(new Error());
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 500 when the app does not exit for tenant', async () => {
    mockGet.mockImplementationOnce(() => ({
      promise: () => ({}),
    }));
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 500 when there is a error fetching the app in DynamoDB', async () => {
    mockGet.mockRejectedValueOnce(new Error());
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 500 when there is a error validating digital channels credentials', async () => {
    mockGetSecretValue.mockImplementationOnce(() => ({
      promise: () => ({}),
    }));
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 500 when there is error fetching web integration', async () => {
    mockGetIntegrations.mockRejectedValueOnce(new Error());
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });
});
