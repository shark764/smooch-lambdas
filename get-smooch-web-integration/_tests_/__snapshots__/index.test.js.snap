// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get-smooch-web-integration Everything is successful Walkthrough passes in the correct arguments to SmoochCore 1`] = `
Array [
  Array [
    Object {
      "keyId": undefined,
      "scope": "app",
      "secret": undefined,
      "serviceUrl": "mock-smooch-api-url",
    },
  ],
]
`;

exports[`get-smooch-web-integration Everything is successful Walkthrough passes in the correct arguments to decClient.get() 1`] = `
Array [
  Array [
    Object {
      "Key": Object {
        "id": "667802d8-2260-436c-958a-2ee0f71f73f0",
        "tenant-id": "66d83870-30df-4a3b-8801-59edff162034",
      },
      "TableName": "us-east-1-dev-smooch",
    },
  ],
]
`;

exports[`get-smooch-web-integration Everything is successful Walkthrough passes in the correct arguments to secretsClient.getSecretValue() 1`] = `
Array [
  Array [
    Object {
      "SecretId": "us-east-1-dev-smooch-app",
    },
  ],
]
`;

exports[`get-smooch-web-integration Everything is successful Walkthrough passes in the correct arguments to smooch.integrations.get() 1`] = `
Array [
  Array [
    undefined,
    "667802d8-2260-436c-958a-2ee0f71f73f0",
  ],
]
`;

exports[`get-smooch-web-integration Everything is successful Walkthrough passes in the correct arguments to validatePlatformPermissions 1`] = `
Array [
  Array [
    Object {
      "user-id": "667802d8-2260-436c-958a-2ee0f71f73f0",
    },
    Array [
      "PLATFORM_VIEW_ALL",
    ],
  ],
]
`;

exports[`get-smooch-web-integration Everything is successful Walkthrough passes in the correct arguments to validateTenantPermissions 1`] = `
Array [
  Array [
    "66d83870-30df-4a3b-8801-59edff162034",
    Object {
      "user-id": "667802d8-2260-436c-958a-2ee0f71f73f0",
    },
    Array [
      "WEB_INTEGRATIONS_APP_READ",
    ],
  ],
]
`;

exports[`get-smooch-web-integration Everything is successful sends back status 200 if the code runs without error 1`] = `
Object {
  "body": Object {
    "result": Object {
      "0": Object {
        "app-id": "5e31c81640a22c000f5d7f28",
        "type": "",
      },
      "actionColor": "#1be54c",
      "brandColor": "#1b3de5",
      "conversationColor": "#e53f1b",
      "prechatCapture": "name",
      "whitelistedUrls": Array [
        "url1",
        "url2",
      ],
    },
  },
  "status": 200,
}
`;

exports[`get-smooch-web-integration Everything is successful when brandcolor, conversationcolor and actioncolor is not provided 1`] = `
Object {
  "body": Object {
    "result": Object {
      "0": Object {
        "app-id": "5e31c81640a22c000f5d7f28",
        "type": "",
      },
      "prechatCapture": "name",
      "whitelistedUrls": Array [
        "url1",
        "url2",
      ],
    },
  },
  "status": 200,
}
`;

exports[`get-smooch-web-integration Everything is successful when prechatCapture is disabled 1`] = `
Object {
  "body": Object {
    "result": Object {
      "0": Object {
        "app-id": "5e31c81640a22c000f5d7f28",
        "type": "",
      },
      "prechatCapture": "none",
      "whitelistedUrls": undefined,
    },
  },
  "status": 200,
}
`;

exports[`get-smooch-web-integration sends back status 400 error when there are invalid parameters 1`] = `
Object {
  "body": Object {
    "message": "Error: invalid params value \\"id\\" is required",
  },
  "status": 400,
}
`;

exports[`get-smooch-web-integration sends back status 400 error when there are not enough permissions 1`] = `
Object {
  "body": Object {
    "message": "Error not enough permissions",
  },
  "status": 400,
}
`;

exports[`get-smooch-web-integration sends back status 500 when the app does not exit for tenant 1`] = `
Object {
  "body": Object {
    "message": "The app does not exist for this tenant",
  },
  "status": 500,
}
`;

exports[`get-smooch-web-integration sends back status 500 when there is a error fetching the app in DynamoDB 1`] = `
Object {
  "body": Object {
    "message": "An Error has occurred trying to fetch an app in DynamoDB",
  },
  "status": 500,
}
`;

exports[`get-smooch-web-integration sends back status 500 when there is a error retrieving digital channels credentials  1`] = `
Object {
  "body": Object {
    "message": "An Error has occurred trying to retrieve digital channels credentials",
  },
  "status": 500,
}
`;

exports[`get-smooch-web-integration sends back status 500 when there is a error validating digital channels credentials 1`] = `
Object {
  "body": Object {
    "message": "An Error has occurred trying to validate digital channels credentials",
  },
  "status": 500,
}
`;

exports[`get-smooch-web-integration sends back status 500 when there is error fetching web integration 1`] = `
Object {
  "body": Object {
    "message": "An Error has occurred trying to fetch a web integration",
  },
  "status": 500,
}
`;
